### PR TITLE
KAFKA-14133: Migrate consumer mock in TaskManagerTest to Mockito

### PR DESCRIPTION
**Pull request by clolov**
This pull request migrates the consumer mock in TaskManagerTest test by test for easier reviews.

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 1]
```diff
@@ -197,6 +196,8 @@ public class TaskManagerTest {
     @Mock(type = MockType.STRICT)
     private Consumer<byte[], byte[]> consumer;
     @org.mockito.Mock
+    private Consumer<byte[], byte[]> mockitoConsumer;
```
Comment (clolov) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:200]: I ran into quite a lot of problems when I tried migrating the whole mock, so I decided to do the migration test-by-test. This way problems could be flushed out. By introducing this mock and using `setMainConsumer` on a test-by-test basis things are easier to go through (at least in my opinion 😊)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 2]
```diff
@@ -1168,13 +1166,10 @@ public void shouldHandleMultipleRemovedTasksFromStateUpdater() {
...
             .thenReturn(mkSet(taskToRecycle0, taskToRecycle1, taskToClose, taskToUpdateInputPartitions));
         when(stateUpdater.restoresActiveTasks()).thenReturn(true);
-        when(activeTaskCreator.createActiveTaskFromStandby(taskToRecycle1, taskId01Partitions, consumer))
+        when(activeTaskCreator.createActiveTaskFromStandby(taskToRecycle1, taskId01Partitions, mockitoConsumer))
             .thenReturn(convertedTask1);
         when(standbyTaskCreator.createStandbyTaskFromActive(taskToRecycle0, taskId00Partitions))
             .thenReturn(convertedTask0);
-        expect(consumer.assignment()).andReturn(emptySet()).anyTimes();
-        consumer.resume(anyObject());
-        expectLastCall().anyTimes();
```
Comment (clolov) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:1251]: According to Mockito these were unused. I deleted them in EasyMock world and the test still passed. Hence I removed them.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 3]
```diff
@@ -3010,13 +2993,9 @@ public void initializeIfNeeded() {
             }
         };
 
-        consumer.commitSync(Collections.emptyMap());
-        expectLastCall();
-        expect(consumer.assignment()).andReturn(emptySet());
-        consumer.resume(eq(emptySet()));
-        expectLastCall();
```
Comment (clolov) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:3158]: Same as above, Mockito reported these are unused. I added an assertion in Mockito world at the end (...I should probably add one in the first test as well, but I can do this in part 2 of this pull request)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 4]
```diff
@@ -3047,14 +3027,9 @@ public void completeRestoration(final java.util.function.Consumer<Set<TopicParti
...
-        expectLastCall();
```
Comment (clolov) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:3196]: Ditto

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 5]
```diff
@@ -3664,20 +3639,11 @@ public void shouldCloseStandbyTasksOnShutdown() {
...
         when(standbyTaskCreator.createTasks(assignment)).thenReturn(singletonList(task00));
 
-        // `tryToCompleteRestoration`
-        expect(consumer.assignment()).andReturn(emptySet());
-        consumer.resume(eq(emptySet()));
-        expectLastCall();
-
-        // `shutdown`
-        consumer.commitSync(Collections.emptyMap());
-        expectLastCall();
```
Comment (clolov) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:3818]: Ditto

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 6]
```diff
@@ -4809,11 +4768,14 @@ public void shouldNotFailForTimeoutExceptionOnCommitWithEosAlpha() {
             exception.corruptedTasks(),
             equalTo(Collections.singleton(taskId00))
         );
+
+        Mockito.verify(mockitoConsumer, times(2)).groupMetadata();
```
Comment (clolov) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:4928]: Mockito claims the method is called twice. This wasn't specified in EasyMock world, but I decided to make it explicit now.

**Review by clolov — COMMENTED** [review]
This is part 1 of migrating the consumer mock from EasyMock to Mockito. I am envisioning that I will need one more pull request to complete it. I decided to spit them because a) it was difficult for me to not move things test by test and b) I think it is easier to review

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 7]
```diff
@@ -4954,8 +4910,6 @@ public void shouldConvertStandbyTaskToActiveTask() {
         when(activeTaskCreator.createActiveTaskFromStandby(Mockito.eq(standbyTask), Mockito.eq(taskId00Partitions), any()))
             .thenReturn(activeTask);
 
-        replay(consumer);
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:5105]: I assume that consumer is never used in these tests where we have removed the replay?

In that case, should we do something similar to what you did in shouldReturnFalseWhenThereAreStillNonRunningTasks , i.e. test for no interactions

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 6 above)
Comment (divijvaidya) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:4928]: should it be called twice as per code?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 8]
```diff
@@ -1354,14 +1346,12 @@ public void shouldHandleTimeoutExceptionInTransitRestoredTaskToRunning() {
         final TaskManager taskManager = setUpTransitionToRunningOfRestoredTask(task, tasks);
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:1433]: don't we want to do a `taskManager.setMainConsumer(mockitoConsumer);` here? (unless setUpTransitionToRunningOfRestoredTask is already doing so)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 6 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:4928]: I verified that if I change
```
expect(consumer.groupMetadata()).andStubReturn(null);
```
to
```
expect(consumer.groupMetadata()).andReturn(null).times(2);
```
the test still passes. I suspect it was always called twice, just never verified

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 9]
```diff
@@ -4904,7 +4870,8 @@ public void suspend() {
         final Map<TaskId, Set<TopicPartition>> assignment = new HashMap<>(taskId00Assignment);
         assignment.putAll(taskId01Assignment);
         when(activeTaskCreator.createTasks(any(), Mockito.eq(assignment))).thenReturn(asList(task00, task01));
-        replay(consumer);
+
+        taskManager.setMainConsumer(mockitoConsumer);
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:5023]: Is there a place where we are using old consumer with taskManager? If not, can we se this to new consumer in @before methods itself where we construct taskManager

**Review by divijvaidya — COMMENTED** [review]
Left a small comment, otherwise looks good

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 9 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:5023]: Yes, we are using the old consumer in all places making calls to
```
    private static void expectConsumerAssignmentPaused(final Consumer<byte[], byte[]> consumer) {
        final Set<TopicPartition> assignment = singleton(new TopicPartition("assignment", 0));
        expect(consumer.assignment()).andReturn(assignment);
        consumer.pause(assignment);
    }
```
This is also why I took the approach of having a separate Mockito consumer. Part 2 of this pull request will migrate all of those usages.

**Review by divijvaidya — APPROVED** [review]
Looks good!

**Comment by divijvaidya** [comment]
The test modified in this PR is successful in CI - https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15112/5/testReport/org.apache.kafka.streams.processor.internals/TaskManagerTest/

**Pull request merged**